File: scala/jersik-scala-testclient/src/test/scala/io/encoded/jersik/scala/testclient/RoundtripTest.scala
//
// Copyright (c) 2014 Mirko Nasato
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
package io.encoded.jersik.scala.testclient

import java.net.URI
import org.apache.http.impl.client.HttpClients
import org.junit.runner.RunWith
import org.scalatest.FlatSpec
import org.scalatest.Matchers
import io.encoded.jersik.scala.runtime.JsonCodec
import org.scalatest.junit.JUnitRunner
import java.io.FileInputStream
import io.encoded.jersik.scala.runtime.ObjectCodec
import org.scalatest.FunSuite
import io.encoded.jersik.testsuite._

@RunWith(classOf[JUnitRunner])
class RoundtripTest extends FunSuite with Matchers {

  val client = new TestServiceClient(HttpClients.createDefault(), URI.create("http://localhost:8080/test/"), JsonCodec)

  roundtripTest("Empty.json", EmptyCodec, client.testEmpty)
  roundtripTest("ScalarFieldsA.json", ScalarFieldsCodec, client.testScalarFieldsA)
  roundtripTest("ScalarFieldsB.json", ScalarFieldsCodec, client.testScalarFieldsB)
  roundtripTest("ScalarFieldsC.json", ScalarFieldsCodec, client.testScalarFieldsC)
  roundtripTest("OptionalFieldsA.json", OptionalFieldsCodec, client.testOptionalFieldsA)
  roundtripTest("OptionalFieldsB.json", OptionalFieldsCodec, client.testOptionalFieldsB)
  roundtripTest("ListFieldsA.json", ListFieldsCodec, client.testListFieldsA)
  roundtripTest("ListFieldsB.json", ListFieldsCodec, client.testListFieldsB)
  roundtripTest("MapFieldsA.json", MapFieldsCodec, client.testMapFieldsA)
  roundtripTest("MapFieldsB.json", MapFieldsCodec, client.testMapFieldsB)
  roundtripTest("StructFieldsA.json", StructFieldsCodec, client.testStructFieldsA)
  roundtripTest("StructFieldsB.json", StructFieldsCodec, client.testStructFieldsB)

  private def roundtripTest[T](fixture: String, objectCodec: ObjectCodec[T], operation: T => T): Unit = {
    test(fixture) {
      val request = JsonCodec.decode(new FileInputStream("../../test-suite/fixtures/"+ fixture), objectCodec)
      println(fixture +": "+ request)
      val response = operation(request)
      assert(response == request)
    }
  }

}

File: java/jersik-java-testclient/src/test/java/io/encoded/jersik/java/testclient/RoundtripTest.java
package io.encoded.jersik.java.testclient;

import static org.junit.Assert.assertEquals;
import io.encoded.jersik.runtime.JsonCodec;
import io.encoded.jersik.runtime.ObjectCodec;
import io.encoded.jersik.testsuite.*;

import java.io.FileInputStream;
import java.lang.reflect.Method;
import java.net.URI;
import java.util.Arrays;
import java.util.Collection;

import org.apache.http.impl.client.HttpClients;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;

@RunWith(Parameterized.class)
public class RoundtripTest {

    private static class Operation {

        private final TestServiceClient serviceClient;
        private final String operationName;
        private final Class<?> requestClass;

        public Operation(TestServiceClient serviceClient, String operationName, Class<?> requestClass) {
            this.serviceClient = serviceClient;
            this.operationName = operationName;
            this.requestClass = requestClass;
        }

        public Object invoke(Object request) throws Exception {
            Method method = serviceClient.getClass().getMethod(operationName, requestClass);
            return method.invoke(serviceClient, request);
        }
    }

    private static final URI SERVICE_URI = URI.create("http://localhost:8080/test/");
    private static final TestServiceClient SERVICE_CLIENT = new TestServiceClient(HttpClients.createDefault(), SERVICE_URI);

    private final String fixture;
    private final ObjectCodec<?> objectCodec;
    private final Operation operation;


    public RoundtripTest(String fixture, ObjectCodec<?> objectCodec, Operation operation) {
        this.fixture = fixture;
        this.objectCodec = objectCodec;
        this.operation = operation;
    }

    @Parameters
    public static Collection<Object[]> parameters() {
        return Arrays.asList(new Object[][] {
            { "Empty.json", EmptyCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testEmpty", Empty.class) },
            { "ScalarFieldsA.json", ScalarFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testScalarFieldsA", ScalarFields.class) },
            { "ScalarFieldsB.json", ScalarFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testScalarFieldsB", ScalarFields.class) },
            { "ScalarFieldsC.json", ScalarFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testScalarFieldsC", ScalarFields.class) },
            { "OptionalFieldsA.json", OptionalFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testOptionalFieldsA", OptionalFields.class) },
            { "OptionalFieldsB.json", OptionalFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testOptionalFieldsB", OptionalFields.class) },
            { "ListFieldsA.json", ListFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testListFieldsA", ListFields.class) },
            { "ListFieldsB.json", ListFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testListFieldsB", ListFields.class) },
            { "MapFieldsA.json", MapFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testMapFieldsA", MapFields.class) },
            { "MapFieldsB.json", MapFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testMapFieldsB", MapFields.class) },
            { "StructFieldsA.json", StructFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testStructFieldsA", StructFields.class) },
            { "StructFieldsB.json", StructFieldsCodec.INSTANCE, new Operation(SERVICE_CLIENT, "testStructFieldsB", StructFields.class) }
        });
    }

    @Test
    public void roundtripTest() throws Exception {
        String fileName = "../../test-suite/fixtures/"+ fixture;
        Object request = JsonCodec.INSTANCE.decode(new FileInputStream(fileName), objectCodec);
        Object response = operation.invoke(request);
        assertEquals(request, response);        
    }

}
